samtranslator/validator/validator.py

```diff
@@ -7,49 +9,254 @@
-        :param dict schema: Optional, dictionary containing JSON Schema representing SAM template
-        :return: Empty string if there are no validation errors in template
+    def _validate(self, template_dict):
         """
+        Validates a SAM Template
 
-        if not schema:
-            schema = SamTemplateValidator._read_schema()
+        Parameters
+        ----------
+        template_dict : dict
+            Template to validate
+        schema : str, optional
+            Schema content, by default None
 
-        validation_errors = ""
+        Returns
+        -------
+        list[str]
+            List of validation errors if any, empty otherwise
+        """
 
-        try:
-            jsonschema.validate(template_dict, schema)
-        except ValidationError as ex:
-            # Stringifying the exception will give us useful error message
-            validation_errors = str(ex)
-            # Swallowing expected exception here as our caller is expecting validation errors and
-            # not the valiation exception itself
-            pass
+        # Tree of Error objects
+        # Each object can have a list of child errors in its Context attribute
+        validation_errors = self.validator.iter_errors(template_dict)
 
-        return validation_errors
+        # Set of "[/Path/To/Element] Error message"
+        # To track error uniqueness, Dict instead of List, for speed
+        errors_set = {}
 
-    @staticmethod
-    def _read_schema():
+        for e in validation_errors:
+            self._process_error(e, errors_set)
+
+        # To be consistent across python versions 2 and 3, we have to sort the final result
+        # It seems that the validator is not receiving the properties in the same order between python 2 and 3
+        # It thus returns errors in a different order
+        return sorted(errors_set.keys())
+
+    def _process_error(self, error, errors_set):
         """
-        Reads the JSON Schema at given file path
+        Processes the validation errors recursively
+        error is actually a tree of errors
+        Each error can have a list of child errors in its 'context' attribute
 
-        :param string schema_file: Optional path to the schema file. If not provided, the system configured value
-            will be used
-        :return dict: JSON Schema of the policy template
+        Parameters
+        ----------
+        error : Error
+            Error at the head
+        errors_set : Dict
+            Set of formatted errors
         """
-        return SamTemplateValidator._read_json(sam_schema.SCHEMA_FILE)
+        if error is None:
+            return
 
-    @staticmethod
-    def _read_json(filepath):
+        if not error.context:
+            # We only display the leaves
+            # Format the message:
+            # [/Path/To/Element] Error message
+            error_path = "/".join([str(p) for p in error.absolute_path]) if error.absolute_path else "/"
```

Comment:
Should this be `.` instead of `/`? I assuming this is for paths in the template? I think its more consistent to use `.` instead of `/`